File: texts/get.js
import _ from 'lodash';
import { select, in as $in, like } from 'sql-bricks';
import table from './lib/table';
import parse, { parseSQLError } from './lib/parse';
import getConnection from './lib/getConnection';

export default async (e, context, callback) => {
  const conn = await getConnection();
  const params = _.get(e, 'queryStringParameters') || {};
  const { text, translation } = table;
  const {
    q,
    key,
    lcid,
  } = params;

  let data = [];
  let sql = select()
    .from(text.name)
    .join(translation.name, { [translation.columns.textId]: text.columns.id })
    .orderBy(`${text.columns.createdAt} DESC`);

  if (!_.isNil(q)) {
    sql = sql.select(text.columns.id).where(like(translation.columns.body, `%${q}%`)).groupBy(text.columns.id);
    sql = select()
      .from(text.name)
      .join(translation.name, { [translation.columns.textId]: text.columns.id })
      .where($in(text.columns.id, sql))
      .orderBy(`${text.columns.createdAt} DESC`);
  }

  if (!_.isNil(key)) {
    sql = sql.where(like(text.columns.key, `%${key}%`));
  }

  if (!_.isNil(lcid)) {
    sql = sql.where(translation.columns.lcid, lcid);
  }

  let response;
  const headers = { 'Access-Control-Allow-Origin': '*' };

  try {
    const grouped = _.groupBy(await conn.query(sql.toString()), 'textId');
    data = _.map(_.values(grouped), t => parse(t, true));

    response = {
      body: JSON.stringify({ data }),
      headers,
      statusCode: 200,
    };
  } catch (error) {
    response = {
      body: JSON.stringify({ error: parseSQLError(error) }),
      headers,
      statusCode: 400,
    };
  } finally {
    conn.end();
  }

  callback(null, response);
};

File: works/get.py
from contrib import db, dumps, split
from orator.query.join_clause import JoinClause


def handler(event, context):
    sql = db.table('work')
    query_params = event.get('queryStringParameters') or {}

    authors = split(query_params.get('authors'))
    categories = split(query_params.get('categories'))

    if authors:
        sql = (
            sql.join('workRole', 'work.id', '=', 'workRole.workId')
            .where('workRole.type', 'author')
            .where_in('workRole.userId', authors)
        )

    if categories:
        sql = (
            sql.join(
                JoinClause('categoryMap')
                .on('work.id', '=', 'categoryMap.categorizableId')
                .where('categoryMap.categorizableType', '=', 'work')
            )
            .join('category', 'category.id', '=', 'categoryMap.categoryId')
            .where_in('category.key', categories)
        )

    return {
        'body': dumps({'data': sql.get()}),
        'headers': {'Access-Control-Allow-Origin': '*'},
        'statusCode': 200,
    }
